### ISSUE TITLE
OCM spec to be extend to reference Hashicorp Vault KV records as a resource

### ISSUE THREAD

**Issue by robertwol** [open-component-model/ocm-spec#83]
# Placeholder to be refined.

**What would you like to be added**:
This issue is about extending the OCM Spec to include the referencing to the Hashicorp Vault KV records (secrets) in similar fashion to Git

**Why is this needed**:
This allow to referecen resources located within Vault

@mandelsoft @sidhishruti22 could you help with providing more reasonable description and specific requirements.

**Comment by morri-son** [comment]
@robertwol , what should be the purpose of this request? Can you give an example? Vault is used to store content in an encrypted format. Referencing such content in an OCM descriptor would mean that access to vault is required, which means passwords, tokens or certificates have to be specified. Or is this issue about adding Hashicorps vault as an identity provider in the .ocmconfig file, so that credentials for accessing OCM repositories can be stored in the Vault and the configuration points to that?

**Comment by jensh007** [comment]
I agree that an example would be helpful to explain the use case for this request.

**Comment by sidhishruti22** [comment]
As of now, we are using "github_username" and "github_token"(git_pat) for our use-case . 
and we are fetching it from the vault.

**Comment by jensh007** [comment]
@sidhishruti22 Can you please give more details:

- Why should this be part of the spec and not just a feature of the CLI?
- How do you propose to extend the spec: Please give an example of how a component-descriptor should look like and at what level you want to have this extension.
-

**Comment by sidhishruti22** [comment]
hey @jensh007 , its on low priority right now, will share the details when we have some more information.
cc @robertwol

**Comment by mandelsoft** [comment]
THis is not part of the OCM spec, it is even not part of the ocm model.

The OCM library provides an own mechanism to provide credentials to be accessed by OCM extensions.

Therefore, an own Context type `crednetials` is used. Like the other context types (and the OCM content type)
it can be extended by (here credential) repository types. So far, there is the dockerconfig repo type. What would have to be done is to implement a new kind of repository providing access to credentials stored in a vault.

The credentials provisioning in the OCM lib then consist of four parts:
- the *credential repositories*
  a credential repository is an abstraction for a set of names credential sets (for example username and password)..
- the notion of a *consumer identity*
  This is just a set of attributes plus a type attribute. Every kind of consumer, for example an OCI registry, or a GitHub repository
  defines an on type key and set of properties defining the environment credentials should be used for (for example, the GitHub server URL and the GitHub Repository, or the OCI registry URL and a repository path prefix, etc)
- configurable mapping, mapping consumer ids to credentials in credential repositories.
- Potentially type specific consumer matchers.
  It matches a given credential request, consisting of the consumer id for a dedicated technical environment against the configured credential mappings. The best matching mapping is used to get the credentials from the mapped credential repository. For example this matching might just be a one to one matching of the consumer attributes, or more complex. For example to match a given repository path in the request by a path prefix given in the crednential mapping. 

A piece of code requiring credentials for a dedicated environment then composes the consumer id for this environment according to the rules for the dedicated type of environment (for example OCI or GitHub). and asks the credential context to provide credentials for this id. Here, the mappings are matched and the credentials are taken from the found credential set (in the mapped credentials repository)

Via configuration the credential context can be configured in a generic manner to know various credential repositories and and dedicated consumer id mapping.
